Remove deprecated torch.solve

IvanYashchuk opened the pull request:
The time has come to remove deprecated linear algebra related functions. This PR removes `torch.solve`.

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano

IvanYashchuk committed:
Remove deprecated torch.solve

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/viable/strict' into remove-solve

Lezcano commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -278,22 +270,6 @@ template<class scalar_t, class value_t=scalar_t>
-}
-
-template<> void lapackSolve<float>(int n, int nrhs, float *a, int lda, int *ipiv, float *b, int ldb, int *info) {
-  sgesv_(&n, &nrhs, a, &lda, ipiv, b, &ldb, info);
```
Why don't we want to keep these kernels? Shouldn't they be more efficient than lu + lu_solve?

IvanYashchuk replied:
No, they are not more efficient. `gesv` = `getrf` + `getrs` but done in an external library. We should leave only the code that is used.

Lezcano commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cpp:
```diff
@@ -163,86 +153,6 @@ void magmaGels(
     scalar_t* dA, magma_int_t ldda, scalar_t* dB, magma_int_t lddb,
     scalar_t* hwork, magma_int_t lwork, magma_int_t* info);
 
-template<>
-void magmaSolve<double>(
-    magma_int_t n, magma_int_t nrhs, double* dA, magma_int_t ldda,
-    magma_int_t* ipiv, double* dB, magma_int_t lddb, magma_int_t* info) {
-  MagmaStreamSyncGuard guard;
-  magma_dgesv_gpu(n, nrhs, dA, ldda, ipiv, dB, lddb, info);
-  AT_CUDA_CHECK(cudaGetLastError());
-}
```
Same question for these.

lezcano reviewed (commented):
I won't cry when this one's gone either. Now, I left a few comments regarding the MAGMA / LAPACK functions that we're removing.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into remove-solve

mruberry commented:
Build issues look real here, too

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/viable/strict' into remove-solve

IvanYashchuk committed:
Remove traces of _solve_helper_cuda in lazy linalg dispatcher

IvanYashchuk commented:
I resolved the build issues.

lezcano reviewed (approved):
Happy with this one once the CI is green.

I will resolve the merge conflicts in https://github.com/pytorch/pytorch/pull/74046 when this gets merged. cba to roll them back now really.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/viable/strict' into remove-solve

IvanYashchuk committed:
Merge branch 'master' into remove-solve

albanD reviewed (commented):
Thanks for the PR Ivan.
Do you think it would be possible to add to `torch/__init__.py` something like:
```python
# This function was depreacted and removed.
# This nice error message can be removed in version 1.3+
def solve(input, A, out=None):
    raise RuntimeError("This function was deprecated since version 1.9 and is now removed. Please use the `torch.linalg.solve` function instead")
```

To make sure that users get a nice error message for a couple versions?

IvanYashchuk commented:
Sure, I will add that.

IvanYashchuk committed:
Remove include <ATen/ops/_solve_helper_native.h>

IvanYashchuk committed:
Add function and method with an error about the removal

IvanYashchuk commented:
Alright, I added a `solve` function and a method that raises an error:
```py
In [1]: import torch
In [2]: a = torch.randn(3, 3)
In [3]: b = torch.randn(3, 3)
In [4]: torch.solve(b, a)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-4-6f9b9974b5b1> in <module>
----> 1 torch.solve(b, a)

RuntimeError: This function was deprecated since version 1.9 and is now removed. Please use the `torch.linalg.solve` function instead.

In [5]: b.solve(a)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-5-fd669f2275c3> in <module>
----> 1 b.solve(a)

RuntimeError: This function was deprecated since version 1.9 and is now removed. Please use the `torch.linalg.solve` function instead.
```

Lezcano commented:
If you ever manage to land this, remind me to remove in https://github.com/pytorch/pytorch/pull/74046 the backward functions for `torch.solve`. Big if though.

IvanYashchuk committed:
Merge branch 'master' into remove-solve

IvanYashchuk committed:
Add Tensor.solve to torch._overrides.get_ignored_functions

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/viable/strict' into remove-solve

IvanYashchuk committed:
Merge branch 'master' into remove-solve

albanD commented on test/test_linalg.py:
```diff
@@ -6351,12 +6264,10 @@ def test_solve_methods_arg_device(self, device):
             # triangular_solve goes through specific backend dispatch (CPU/CUDA) and hit auto-generated device check first
             generic_backend_dispatch_err_str = "Expected b and A to be on the same device"
             specific_backend_dispatch_err_str = "Expected all tensors to be on the same device"
-            with self.assertRaisesRegex(RuntimeError, generic_backend_dispatch_err_str):
```
nit: could you leave a small test here (or in test_torch.py) that the error you added is working as expected?

IvanYashchuk replied:
Sure, I added the test in https://github.com/pytorch/pytorch/pull/70986/commits/68928e736f23e2d922023ac84f5f2622a4e338ed.

albanD reviewed (commented):
Sounds good!

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/viable/strict' into remove-solve

IvanYashchuk committed:
Add test for raising an error when solve is called

IvanYashchuk committed:
Move 'def solve' to avoid merge conflict

IvanYashchuk commented:
I added a test checking that the error is raised as expected and resolved accumulated merge conflicts.

albanD reviewed (approved):
Sounds good!

albanD commented:
@pytorchbot merge this please
